### PR TITLE
47 multiple minor bugs that arose trying to get training to run

### PR DESCRIPTION
**Pull request by nelson-lojo**
yes I know there's a typo

### CODE REVIEW

**Review by nelson-lojo — COMMENTED** [review]
Summary of changes:
* proper python types in yaml config files
* reordered models in models.yml to (hopefully) improve readability
* Added 1 step to training runs to save last checkpoint
* reduced checkpoint frequency by a factor of 10 so we don't run out of wandb space
* commented out baseline models in training runs to speed up training
* Added pytorch-cuda drivers to dependencies to work with bigger GPUs
* moved yaml logging into wandb initialization
* added cli options to pass project name and run name into wandb
* fixed sparse linear function class to sample weight dimensions to keep from dimensions in the curriculum
* some changes to GDModel ContextModel to enable error-less execution (potentially some correctness issues)
    * detach some tensors to reduce memory usage (need to check if this breaks predictions)
    * changed batch size handling
    * y shapes adjusted
    * predictions and other things moved to coda if available
* Sparse linear ContextModel changes
    * predictions and parameters moved to cuda if available
    * y shapes adjusted
    * Added some handling for convergence warnings (this may be very frivolous/redundant)
* KNN ContextModel changes
    * moved params and predictions to cuda if available
    * adjusted y dimensions
* Llama now uses the same MLP width as what GPT2 would
* Don't skip the first curriculum stage (thanks @Famara72)
* Actually change the sequence length along the curriculum (thanks @Famara72)
* Run baseline models with `model.eval()` instead of `torch.no_grad()`

File: src/models/linear.py [hunk 1]
```diff
@@ -108,7 +109,10 @@ def forward(self, xs, ys):
                     with warnings.catch_warnings():
                         warnings.filterwarnings("error")
                         try:
-                            clf.fit(train_xs.numpy(), train_ys.numpy())
+                            clf.fit(train_xs, train_ys)
```
Comment (riensou) [inline src/models/linear.py:112]: I get this warning, are we okay with this?

Argument of type "Tensor" cannot be assigned to parameter "X" of type "MatrixLike | coo_matrix" in function "fit"
  Type "Tensor" cannot be assigned to type "MatrixLike | coo_matrix"
    "Tensor" is incompatible with "coo_matrix"
    "Tensor" is incompatible with "ndarray[Unknown, Unknown]"
    "Tensor" is incompatible with "DataFrame"
    "Tensor" is incompatible with "spmatrix"Pylance[reportArgumentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportArgumentType)

File: conf/models.yml [hunk 2]
```diff
@@ -57,12 +50,9 @@ model: &mlp
         dimensions: [100]
     opt_alg_name: adam
     batch_size: 100
-    lr: 5e-3
+    lr: !!float 5e-3
```
Comment (riensou) [inline conf/models.yml:53]: [yaml documentation](https://pyyaml.org/wiki/PyYAMLDocumentation#yaml-tags-and-python-types)

is there a reason we want to specify the float type here, but not the other types (int, str, etc.)?

File: src/__main__.py [hunk 3]
```diff
@@ -75,5 +83,7 @@ def main(args: arg.Namespace):
                         help="path to the config file to use for training")
     parser.add_argument("--resume", '-r', type=str, default="", action='store', dest="checkpointfile",
                         help="path to the checkpoint to resume training with")
+    parser.add_argument("--wandb-project", type=str, default="", dest="projectname")
+    parser.add_argument("--run-name", type=str, default="", dest="runname")
```
Comment (riensou) [inline src/__main__.py:87]: should we add `help` information to each of these arguments as the other ones have? seems self-explanatory enough probably anyways

File: src/models/nearest_neighbors.py [hunk 4]
```diff
@@ -16,7 +16,7 @@ def forward(self, xs, ys):
 
         for i in range(ys.shape[1]):
             if i == 0:
-                preds.append(torch.zeros_like(ys[:, 0]))  # predict zero for first point
+                preds.append(torch.zeros(xs.shape[:1] + torch.Size([1, self.y_dim]), device=xs.device))  # predict zero for first point
```
Comment (riensou) [inline src/models/nearest_neighbors.py:19]: is the 1 here so that it can be broadcasted, or am I misunderstanding how this line works?

File: src/train/context_trainer.py [hunk 5]
```diff
@@ -85,17 +85,18 @@ def train(self, pbar: Optional[Any] = None) -> ContextModel:
...
-                        baseline_output = baseline(x_batch, y_batch)
-
-                        if baseline_output.shape != y_batch.shape:
-                            raise ValueError(
-                                f"Baseline model {baseline.name} produced ill-shaped predictions!" + \
-                                    f"Expected: {y_batch.shape}    Got: {baseline_output.shape}"
-                            )
-
-                        baseline_loss[baseline.name] = self.loss_fn(baseline_output.cpu(), y_batch.cpu())
+                # with torch.no_grad():
```
Comment (riensou) [inline src/train/context_trainer.py:88]: delete this comment

**Review by riensou — APPROVED** [review]
looks good, some small changes / clarifications

File: conf/models.yml (hunk 2 above)
Comment (nelson-lojo) [reply conf/models.yml:53]: the value here defaults to a string (because of the "e"), so we cast to a float for torch's optimizers

File: src/models/nearest_neighbors.py (hunk 4 above)
Comment (nelson-lojo) [reply src/models/nearest_neighbors.py:19]: This is for `torch.cat` later down. We need the dimension present to concatenate with predictions later on. This line predicts a constant 0 for the first point in every sequence.

File: src/__main__.py (hunk 3 above)
Comment (nelson-lojo) [reply src/__main__.py:87]: done

File: src/train/context_trainer.py (hunk 5 above)
Comment (nelson-lojo) [reply src/train/context_trainer.py:88]: done

File: src/models/linear.py (hunk 1 above)
Comment (nelson-lojo) [reply src/models/linear.py:112]: Converted to numpy arrays, hopefully this makes all errors raise properly. Device moving shouldn't matter since sklearn only works on cpu anyway.

**Review by riensou — APPROVED** [review]
ready to merge
